### ISSUE TITLE
flash messages (put_flash/3) not closing

### ISSUE THREAD

**Issue by MarthinL** [phoenixframework/phoenix#5470]
Hi guys,

It's probably something small, but it's in an area of Phoenix that's largely invisible to mere mortals like me. The flash messages not disappearing by themselves, and not responding to clicking on the close button caught my attention while testing a fresh project's phx.gen.auth code. But I did more tests and the problem has nothing to do with the generated auth code as the behaviour is exactly the same in a bare-bones freshly generated project.

### Recreating the problem
Generate a new project, e.g.

```
mix phx.new test
cd test
mix ecto.create
```

Then edit the home page at the controller level to insert a put_flash call to the conn, e.g. lib/test_web/controllers/page_controller.ex

```
  def home(conn, _params) do
    # The home page is often custom made,
    # so skip the default app layout.
    conn
    |> put_flash(:info, "Home is where the heart is")
    |> render(:home, layout: false)
  end

```

Finally, run the server and open the home page in the browser.

```
iex -S mix phx.server
```

### Environment

* Elixir version (elixir -v): Elixir 1.14.5 (compiled with Erlang/OTP 25)
* Phoenix version (mix deps): {:phoenix, "~> 1.7.2"}
* Operating system: macOS

### Expected behavior

As expected, put_flash formats and populates pre-defined area of the page with the key and message parameters as provided. The message area is then expected to fade away (ease-in transition) after a set amount of time. There is also a close icon (button) displayed on the message area by which the user can manually close the area if they wish to do so rather than wait for it to disappear.

### Actual behavior

The displayed flash message area stays visible indefinitely even if the close button is clicked.

### Additional notes

*  In the code to reproduce the error as simply as possible, reloading the page would call put_flash again and it would display again. I can confirm from the original testing I've done with auth UI, reloading the page is the only way to get rid of the flash message which obscures part of the page.

* I've run the same test and got the same result in all the mainstream browsers.

* I had a virtually empty project on my computer that was generated with phoenix v1.17.1 and I ran the same test there. It might have been because of something else that was out of sync, but the result I got in that test was that put_flash failed to render the flash message area visible at all.

I noticed that there had been a recent issue with flash messages not appearing as they should (Issue #5247), so by the "law" of blaming the most recent changes it seems possible that this had been an unintended consequence of that work. Unfortunately it is well above my pay grade in terms of javascript, how phoenix generates code and how that code interacts with the javascript environment to debug the code or even know where to look for the code that may be involved and might have changed recently.

**Comment by phoebe100** [comment]
Hi,
how do you start your application and which phoenix version (you mentioned both 1.7.1 and 1.7.2) are you using?

For me, the close button works fine.
For comparison, here are my specs:

- Elixir 1.14.5 (compiled with Erlang/OTP 25)
- Phoenix v1.7.2
- Application start with mix phx.server (directly on my machine, without docker or any container wrapping)

**Comment by MarthinL** [comment]
If you are familiar with the code involved and how it's getting generated, perhaps you can point out where I could point a debugger or insert a log statement to check whether the code that is meant to run is not getting called or failing to do what it's meant to do? 

The syntax of the phx-click attribute on the flash message div itself is unfamiliar to me, but the aim seems to be to result in a series of style changes to the div. In my environment I was unable to detect any values to those style properties being set on the actual div, before or after pressing the button. That might be a clue. For starters, what code should run (in what file, what function) on phx-click as specified on that div with id="flash"? I can try to set a breakpoint in the browser debugger and take it from there.

At the moment I have nothing to go on except that it doesn't behave as expected in my environment which to the best of my knowledge is as it would be after upgrading to 1.7.2.

Perhaps one our environments are not "pure" 1.7.2. That in itself would be a major concern since it would imply that the upgrade procedures are incomplete. Either way, getting to the bottom of why this is happening involves isolating the cause of the problem I've reported. For that, someone that knows how it's meant to work should cast their knowing eyes on what is actually happening to spot the difference. I'm hoping that's you, so just tell me what to do to help you see what is happening in my environment. 


> Hi, how do you start your application and which phoenix version (you mentioned both 1.7.1 and 1.7.2) are you using?

I start the app using mix phx.server and use, as stated, Phoenix 1.7.2.

I've also restarted my machine, cleared browser caches and ran using iex -S mix phx.server, and it made no difference.


> 
> For me, the close button works fine. For comparison, here are my specs:
> 
> * Elixir 1.14.5 (compiled with Erlang/OTP 25)
> * Phoenix v1.7.2
> * Application start with mix phx.server (directly on my machine, without docker or any container wrapping)

**Comment by MarthinL** [comment]
@phoebe100, can I ask you this. Does the flash message disappear by itself after a while in your environment? How long? The phx-click code has duration-200 in it, but I don't know if that's milliseconds, seconds or something else.

Also, the flash message I get on the screen (at least by the time I can see it in the debugger appears to be a zombie, i.e. it's dead in the conventional sense of the word but it still displays. I've noticed in the phoenix 1.7.0 and 1.7.1 version I tested with (since your last message) that it is missing the class (in the tailwindcss nomenclature) of hidden. Adding that class in the debugger hides it as would be expected. I also see that the parameters of the pxh-click call refers to "hide" and not to "hidden". If all those parameters are in fact tailwind classes, that "hide" should probably be "hidden" to work correctly, but that's gross speculation. Just trying to be helpful.

**Comment by MarthinL** [comment]
I did the test on Windows (I'm normally using macOS as stated)  and found that the flash box does close on the Windows version. The flash message on neither version disappears after 200 anythings.

@phoebe100, can you confirm that your testing environment was Windows or at least not macOS?

I've run a diff between the code generated on Windows and the code generated on macOS. There are many differences that appear legit because of the differences in operating environment, but also some changes that looks likely to be causing the misbehaviour. I was under the impression that the macOS and Windows elixir and javascript code would be identical, but it does not seem to be the case.

What am I missing?

**Comment by chrismccord** [comment]
The flash messages do not automatically disappear. You can see an example here of how to do that with a js hook:
https://github.com/fly-apps/live_beats/blob/712204ca172ed6d36456c8ed4fd6aaba50099447/assets/js/app.js#LL28C1-L39C2
Thanks!

**Comment by henriquegasques** [comment]
Sorry, I know this issue is closed, but I'm having the same problem. If you don't mind @MarthinL could you please provide more information on how did you solve it?

**Comment by MarthinL** [comment]
There was nothing to solve, actually. 

For the most part I was wrong to expect flash messages to disappear by themselves. I misinterpreted the duration parameter as how long it should remain visible before closing by themselves, rather than what it actually means which is how long (in milliseconds) it should take to transition to visible.

The malfunction that went with it was a result of an old version of the phoenix javascript module lurking around on my development machine from an abandoned older project where I accidently installed it for global use. Once I cleaned that up so the correct code ran, the messages worked as they should.
